data-prepper-plugins/geoip-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/GeoIPProcessorService.java

```diff
@@ -5,44 +5,137 @@
+    private List<DatabasePathURLConfig> databasePath;
+    private final String tempPath;
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final DBSourceOptions dbSourceOptions;
+    public static volatile boolean downloadReady;
+    private boolean toggle;
+    private String flipDatabase;
+
+
     /**
      * GeoIPProcessorService constructor for initialization of required attributes
      * @param geoIPProcessorConfig geoIPProcessorConfig
      * @param tempPath tempPath
      */
     public GeoIPProcessorService(GeoIPProcessorConfig geoIPProcessorConfig, String tempPath) {
-       //TODO
+        this.toggle = false;
+        this.geoIPProcessorConfig = geoIPProcessorConfig;
+        this.tempPath = tempPath;
+        this.databasePath = geoIPProcessorConfig.getServiceType().getMaxMindService().getDatabasePath();
+        flipDatabase = DATABASE_1;
+
+        dbSourceOptions = DbSourceIdentification.getDatabasePathType(databasePath);
+        final Duration checkInterval = Objects.requireNonNull(geoIPProcessorConfig.getServiceType().getMaxMindService().getCacheRefreshSchedule());
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+        scheduledExecutorService
+                .scheduleAtFixedRate(this::downloadThroughURLandS3, 0L, checkInterval.toSeconds(), TimeUnit.SECONDS);
+
+        synchronized (this) {
+            try {
+                while (!downloadReady) {
+                    wait();
+                }
+            } catch (InterruptedException ex) {
+                LOG.info("InterruptedException {0} ",  ex);
+                Thread.currentThread().interrupt();
+            }
+            String finalPath = tempPath + File.separator;
+            licenseType = LicenseTypeCheck.isGeoLite2OrEnterpriseLicense(finalPath.concat(flipDatabase));
+            if (licenseType.equals(LicenseTypeOptions.FREE)) {
+                geoData = new GetGeoLite2Data(finalPath.concat(flipDatabase), geoIPProcessorConfig.getServiceType().getMaxMindService().getCacheSize(), geoIPProcessorConfig);
+            }
+            else if (licenseType.equals(LicenseTypeOptions.ENTERPRISE)) {
+                geoData = new GetGeoIP2Data(finalPath.concat(flipDatabase), geoIPProcessorConfig.getServiceType().getMaxMindService().getCacheSize(), geoIPProcessorConfig);
+            }
+        }
+        downloadReady = false;
     }
 
     /**
-     * Calling downlaod method abased on the database path type
-     * @param DBSourceOptions DBSourceOptions
+     * Calling download method based on the database path type
      */
-    public void downloadThroughURLandS3(DBSourceOptions DBSourceOptions) {
-        //TODO
+    public synchronized void downloadThroughURLandS3() {
+        DBSource dbSource;
+
+        if (!toggle) {
+            flipDatabase = DATABASE_1;
+            toggle = !toggle;
+        } else {
+            flipDatabase = DATABASE_2;
+            toggle = !toggle;
```

Comment:
Since `togge = !toggle` is done in both cases, if can be done outside of if-else block.

Reply:
Moved the line "togge = !toggle" above if and else block.